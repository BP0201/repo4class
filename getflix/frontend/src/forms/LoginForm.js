import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import Alert from "../other/Alert";

function LoginForm({ login }) {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    username: "",
    password: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  // console.debug(
  //     "LoginForm",
  //     "login=", typeof login,
  //     "formData=", formData,
  //     "formErrors", formErrors,
  // );

  // handles form submission
  async function handleSubmit(evt) {
    evt.preventDefault();
    console.log("formData in handleSubmit:", formData)
    if (formData.username.length < 3) {
      setFormErrors(formErrors => [...formErrors, "Username must be between 3-16 characters"])
    }
    if (formData.password.length < 5) {
      setFormErrors(formErrors => [...formErrors, "Password must be between 5-20 characters"])
    }
    let result = await login(formData);
    if (result.success) {
      // redirect if successful
      navigate("/");
    } else {
      // if login fails, render this alert
      setFormErrors(formErrors => [...formErrors, "Invalid username/password"]);
    }
  }

  // handles formData state change when input values change
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(l => ({ ...l, [name]: value }));
  }

  return (
      <div className="LoginForm">
        <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
          <h2 className="my-4">Log In</h2>

          <div className="card shadow">
            <div className="card-body">
              <form onSubmit={handleSubmit}>
                <div className="form-group">
                  <label htmlFor="username">Username</label>
                  <input
                      id="username"
                      name="username"
                      className="form-control"
                      value={formData.username}
                      onChange={handleChange}
                      autoComplete="username"
                      minLength={3}
                      maxLength={16}
                      required
                  />
                </div>
                <div className="form-group">
                  <label htmlFor="password">Password</label>
                  <input
                      id="password"
                      type="password"
                      name="password"
                      className="form-control"
                      value={formData.password}
                      onChange={handleChange}
                      autoComplete="current-password"
                      minLength={5}
                      maxLength={20}
                      required
                  />
                </div>

                {formErrors.length
                    ? <Alert type="danger" messages={formErrors} />
                    : null}

                <button
                    data-testid="login-form-btn"
                    className="btn btn-success float-right"
                    onSubmit={handleSubmit}
                >
                  Login
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
  );
}

export default LoginForm;
