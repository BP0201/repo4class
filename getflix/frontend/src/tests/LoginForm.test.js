import React from 'react';
import { render, fireEvent, screen, act } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import LoginForm from '../forms/LoginForm';

describe("LoginForm", () => {

  // function successfulLogin() {
  //   return { success: true }
  // }

  function failedLogin() {
    return { success: false }
  }

  it("should match the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <LoginForm />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByLabelText("Username")).toBeInTheDocument();
    expect(screen.getByLabelText("Password")).toBeInTheDocument();
  });

  it("should throw an error if inputs are empty", () => {
    render(
      <MemoryRouter>
        <LoginForm login={failedLogin} />
      </MemoryRouter>
    );
    const loginBtn = screen.getByTestId("login-form-btn");
    fireEvent.click(loginBtn);
    expect(screen.getByText("Username must be between 3-16 characters")).toBeInTheDocument();
    expect(screen.getByText("Password must be between 5-20 characters")).toBeInTheDocument();
  });

  // it("should throw an error if user doesn't exist", () => {
  //   render(
  //     <MemoryRouter>
  //       <LoginForm login={failedLogin} />
  //     </MemoryRouter>
  //   );
  //   const usernameInput = screen.getByLabelText("Username");
  //   const passwordInput = screen.getByLabelText("Password");
  //   const loginBtn = screen.getByTestId("login-form-btn");
  //   fireEvent.change(usernameInput, {target: {value: 'testUser'}})
  //   fireEvent.change(passwordInput, {target: {value: 'password1'}})
  //   fireEvent.click(loginBtn);
  //   expect(screen.getByText("Invalid username/password")).toBeInTheDocument();
  // });
});