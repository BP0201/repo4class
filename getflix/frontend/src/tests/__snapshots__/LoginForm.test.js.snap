// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="LoginForm"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h2
        class="my-4"
      >
        Log In
      </h2>
      <div
        class="card shadow"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label
                for="username"
              >
                Username
              </label>
              <input
                autocomplete="username"
                class="form-control"
                id="username"
                maxlength="16"
                minlength="3"
                name="username"
                required=""
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="form-control"
                id="password"
                maxlength="20"
                minlength="5"
                name="password"
                required=""
                type="password"
                value=""
              />
            </div>
            <button
              class="btn btn-success float-right"
              data-testid="login-form-btn"
            >
              Login
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
